File: app/submission.py
"""
이음(EUM) 플랫폼 - 공급자 워크플로우 (L1 적재 확장)
기관 담당자가 업로드한 CSV를 파싱·스키마 추론하여 DuckDB 테이블로
적재하고, 제출(submission) 상태를 관리한다.
"""
import datetime
import uuid

import pandas as pd

from . import database as db

_DTYPE_MAP = {
    "int64": "BIGINT",
    "float64": "DOUBLE",
    "bool": "BOOLEAN",
    "object": "VARCHAR",
    "datetime64[ns]": "TIMESTAMP",
}


def infer_schema(df: pd.DataFrame) -> list[tuple[str, str]]:
    """판다스 DataFrame의 컬럼명·dtype을 DuckDB 컬럼 정의로 변환한다."""
    schema = []
    for col in df.columns:
        duck_type = _DTYPE_MAP.get(str(df[col].dtype), "VARCHAR")
        schema.append((col, duck_type))
    return schema


PREVIEW_LIMIT = 20


def load_csv_to_table(file_obj, table_name: str) -> dict:
    """CSV를 읽어 스키마를 추론하고 DuckDB 테이블로 적재한 뒤 미리보기를 반환한다."""
    df = pd.read_csv(file_obj)
    schema = infer_schema(df)

    conn = db.get_conn()
    conn.execute(f"DROP TABLE IF EXISTS {table_name}")
    conn.register("_upload_df", df)
    conn.execute(f"CREATE TABLE {table_name} AS SELECT * FROM _upload_df")
    conn.unregister("_upload_df")

    preview = db.query(f"SELECT * FROM {table_name} LIMIT {PREVIEW_LIMIT}")
    return {
        "table_name": table_name,
        "rows": len(df),
        "schema": schema,
        "preview": preview,
    }


def new_table_name(tenant_id: str) -> str:
    """제출용 테이블명을 생성한다 (tenant_id + 짧은 uuid로 충돌 방지)."""
    suffix = uuid.uuid4().hex[:8]
    return f"sub_{tenant_id}_{suffix}"


def _now() -> str:
    return datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")


def create_submission(meta: dict, table_name: str, rows: int, quality_summary: str) -> str:
    """메타정보 + 적재 결과 + 진단 요약으로 제출 레코드를 생성한다 (status=submitted)."""
    submission_id = uuid.uuid4().hex
    db.execute(
        "INSERT INTO submissions VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)",
        [
            submission_id, meta["tenant_id"], meta["title"], meta["description"],
            meta["theme"], meta["keywords"], meta["license"], meta["format"],
            table_name, rows, "submitted", quality_summary, None,
            _now(), None,
        ],
    )
    return submission_id


def record_decision(submission_id: str, status: str, decision_note: str = "") -> None:
    """담당자의 승인/반려 결정을 기록한다 (status: approved | rejected)."""
    db.execute(
        "UPDATE submissions SET status = ?, decision_note = ?, decided_at = ? "
        "WHERE submission_id = ?",
        [status, decision_note, _now(), submission_id],
    )


def add_comment(submission_id: str, comment: str) -> str:
    """센터(컨설팅)의 코멘트를 추가한다."""
    comment_id = uuid.uuid4().hex
    db.execute(
        "INSERT INTO consultant_comments VALUES (?,?,?,?)",
        [comment_id, submission_id, comment, _now()],
    )
    return comment_id


def get_submission(submission_id: str) -> dict:
    """제출 상세(메타 + 미리보기 + 코멘트 이력)를 반환한다."""
    meta = db.query("SELECT * FROM submissions WHERE submission_id = ?", [submission_id])[0]
    preview = db.query(f"SELECT * FROM {meta['table_name']} LIMIT {PREVIEW_LIMIT}")
    comments = db.query(
        "SELECT * FROM consultant_comments WHERE submission_id = ? ORDER BY created_at",
        [submission_id],
    )
    return {"meta": meta, "preview": preview, "comments": comments}


def summarize_quality(result: dict) -> str:
    """품질진단 결과를 제출 레코드에 저장할 짧은 요약 문자열로 변환한다."""
    verdict = "통과" if result["passed"] else "미통과"
    return (
        f"규칙 {result['rule_count']}종 / 오류 {result['errors']}건 "
        f"/ 오류율 {result['error_rate']}% / {verdict}"
    )

File: lib/submission.ts
/**
 * 공급자 워크플로우 (app/submission.py → TypeScript)
 * CSV 업로드, 제출 관리, 승인/반려, 코멘트
 */
import type { SupabaseClient } from '@supabase/supabase-js'

const TABLE_NAME_RE = /^sub_(.+)_[0-9a-f]{8}$/

export function validateTableName(name: string): string {
  if (!TABLE_NAME_RE.test(name)) throw new Error(`유효하지 않은 테이블명: ${name}`)
  return name
}

export function newTableName(tenantId: string): string {
  const arr = new Uint8Array(4)
  crypto.getRandomValues(arr)
  const suffix = Array.from(arr).map(b => b.toString(16).padStart(2,'0')).join('')
  return `sub_${tenantId}_${suffix}`
}

export function inferSchema(rows: Record<string, unknown>[]): Array<{ name: string; type: string }> {
  if (!rows.length) return []
  return Object.entries(rows[0]).map(([name, val]) => {
    const type = typeof val === 'number'  ? (Number.isInteger(val) ? 'INTEGER' : 'FLOAT')
               : typeof val === 'boolean' ? 'BOOLEAN'
               : 'TEXT'
    return { name, type }
  })
}

export async function createSubmission(
  supabase: SupabaseClient,
  meta: Record<string, string>,
  tableName: string,
  rows: number,
  qualitySummary: string,
): Promise<string> {
  const arr = new Uint8Array(16)
  crypto.getRandomValues(arr)
  const submissionId = Array.from(arr).map(b => b.toString(16).padStart(2,'0')).join('')
  await supabase.from('submissions').insert({
    submission_id: submissionId,
    tenant_id: meta.tenant_id, title: meta.title, description: meta.description,
    theme: meta.theme, keywords: meta.keywords, license: meta.license, format: meta.format,
    table_name: tableName, rows, status: 'submitted',
    quality_summary: qualitySummary, decision_note: null,
    submitted_at: new Date().toISOString(), decided_at: null,
  })
  return submissionId
}

export async function recordDecision(
  supabase: SupabaseClient,
  submissionId: string,
  status: 'approved' | 'rejected',
  decisionNote = '',
): Promise<void> {
  await supabase.from('submissions').update({
    status,
    decision_note: decisionNote,
    decided_at: new Date().toISOString(),
  }).eq('submission_id', submissionId)
}

export async function addComment(
  supabase: SupabaseClient,
  submissionId: string,
  comment: string,
): Promise<string> {
  const arr = new Uint8Array(16)
  crypto.getRandomValues(arr)
  const commentId = Array.from(arr).map(b => b.toString(16).padStart(2,'0')).join('')
  await supabase.from('consultant_comments').insert({
    comment_id: commentId, submission_id: submissionId,
    comment, created_at: new Date().toISOString(),
  })
  return commentId
}

export async function getSubmission(supabase: SupabaseClient, submissionId: string) {
  const { data: sub } = await supabase
    .from('submissions').select('*').eq('submission_id', submissionId).maybeSingle()
  if (!sub) return null

  const { data: upload } = await supabase
    .from('submission_uploads').select('preview,schema_info')
    .eq('table_name', sub.table_name).maybeSingle()
  const { data: comments } = await supabase
    .from('consultant_comments').select('*')
    .eq('submission_id', submissionId).order('created_at')

  return {
    meta: sub,
    preview: (upload?.preview as unknown[]) ?? [],
    comments: comments ?? [],
  }
}

export function summarizeQuality(result: {
  rule_count: number; errors: number; error_rate: number; passed: boolean
}): string {
  const verdict = result.passed ? '통과' : '미통과'
  return `규칙 ${result.rule_count}종 / 오류 ${result.errors}건 / 오류율 ${result.error_rate}% / ${verdict}`
}
